src/librustc_typeck/collect/type_of.rs

```diff
@@ -399,14 +394,15 @@ fn find_opaque_ty_constraints(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
 
                 // FIXME(oli-obk): trace the actual span from inference to improve errors.
                 let span = self.tcx.def_span(def_id);
-                // used to quickly look up the position of a generic parameter
-                let mut index_map: FxHashMap<ty::ParamTy, usize> = FxHashMap::default();
-                // Skipping binder is ok, since we only use this to find generic parameters and
-                // their positions.
-                for (idx, subst) in substs.iter().enumerate() {
-                    if let GenericArgKind::Type(ty) = subst.unpack() {
+
+                let opaque_generics = self.tcx.generics_of(self.def_id);
```

Comment:
Pre-existing, but this code is in dire need of some comments.

To start, can we add a comment explaining what is happening here -- I think it is this:

---

Examine the type parameters to the opaque type used in the *defining use* and make sure that (a) they are only generic parameters from the enclosing scope and (b) each generic parameter is used at most once. For example, given this function:

```rust
type Foo<A, B> = impl Debug;

fn f1<T, U>() -> Foo<T, U>
```

we would be iterating over the substs `T, U` and checking that they are distinct. We would error for things like `-> Foo<T, T>` or `-> Foo<u32, U>`.

Reply:
`wfcheck` is the primary checker and has a nice comment.
This is just a tombstone insurance policy. But I can try to document it if you want.

Reply:
I'd still like a comment, even if only points over to the "primary comment"

Reply:
In fact, I think pointing out that this is "just" an insurance policy would be quite useful =)